### PR TITLE
MDMP v1.0

### PR DESCRIPTION
**Pull request by AlexKornitzer**
The MDMP plugin is now at a stage where it is usable. It makes use of the PE plugins to analyse the binary contents within the memory dump. This update brings the plugin to the stage where it is feature complete based on the current state/functionality of r2 (0.10.6). Additional functionality such as parsing debugging info is not currently possible (I believe) without changes to r2, which will require a proposal!

As can be seen from the plugin list, MDMP now supports the following:
```                                                                                                                                              
.entries = entries,                                                                                                                                                  
.imports = &imports,                                                                                                                                                  
.info = &info,                                                                                                                                                        
.libs = &libs,                                                                                                                                                                    
.mem = &mem,                                                                                                                                                   
.relocs = &relocs,                                                                                                                                                    
.sections = &sections,                                                                                                                                                
.symbols = &symbols,
```

Additional tests for 32bit and 64bit dumps have been written and will be added to the regression repo.

NOTES:
-Due to the need for PE parsing the PE folder has been cloned into the MDMP folder as we can't link against it while in the extras repo. Once/if MDMP makes it to the main repo this folder can be removed.
-Parsing of additional symbols is not yet possible without refinement to the PE plugin. This is a plugin is in need of cleaning and love (insufficient header guards, memory leaks, etc)!

### CODE REVIEW

**Comment by Maijin** [comment]
Can you modify .travis.yml with your stuff on radare2-regressions see example of m68k https://github.com/radare/radare2-extras/blob/master/.travis.yml#L45

**Comment by AlexKornitzer** [comment]
Done, hopefully correctly.

File: libr/bin/format/mdmp/mdmp.c [hunk 1]
```diff
@@ -8,6 +8,122 @@ ut64 r_bin_mdmp_get_baddr(struct r_bin_mdmp_obj *obj) {
...
+	if (pe_bin) {
+		Pe64_r_bin_pe_free (pe_bin->bin);
+		R_FREE (pe_bin);
+	}
+}
+
+void r_bin_mdmp_free(struct r_bin_mdmp_obj *obj) {
+	if (!obj) return;
+
+	if (obj->streams.ex_threads) r_list_free (obj->streams.ex_threads);
```
Comment (radare) [inline libr/bin/format/mdmp/mdmp.c]: its unnecessary to if() before freeing, because this check is already done in there

**Comment by radare** [comment]
Wow impressive work! i'm merging, the only issue i found was the one i commented you can fix that in another pr later

**Comment by XVilka** [comment]
I think this feature should go into the core repo rather than in extras, since minidumps are in high demand and very common thing to open.

**Comment by AlexKornitzer** [comment]
Thanks guys :) I like XVilka's thinking (not that I am biased or anything!) but MDMP is a new format for me and I am still not familiar with it. Thus, there are a few concerns that I would have about it making its way into the core:

- I could have interpreted certain parts of the protocol incorrectly which could result in errors within the plugin. I have only managed to test it on a small set of data and even then I have only been able to test the aspects that I am familiar with.
- The current implementation is very verbose as it analyses every module in the list, as a result we can end up with massive info lists (i.e. 20k of strings). Would it maybe be more useful to only analyse the executables and ignore the dlls? but would this conflict with other use cases of the MDMP format. At this moment in time I have only been using it to view process memory dumps to search for hooking, reflective loads, etc.

P.S. I'll get on those `if` issues, and I have just noticed a problem with my header guards (nothing major)!

**Comment by radare** [comment]
imho this requires more testing and reviews before merging into master, and obviously not for yesterday’s release

but makes sense to have this in core if we have linux coredump support too.

we should ease the build and install of stuff via r2pm on windows too, maybe distributing precompiled plugins or so

**Comment by AlexKornitzer** [comment]
Yep I totally agree, it would be great if people would test and review it. I am more than happy to be assigned to any issues to do with this plugin when they arrive.

Ah awesome if coredump is in I'll have a peruse through its codebase to see if it can answer some of my outstanding implementation issues.

My mental current roadmap is as follows in the order that requires the least amount of modification to the core:
- Additional refinements to section parsing, will be beneficial to split into .data, .rodata, .text where applicable for a number of reasons, this will also reduce the number of strings (hopefully). I really want to get the memory sections out of the Sections section!
- Remote symbol parsing, i.e. pulling pdb files from the internet if required similar to WinDbg. Requires changes to pe.c which will take me a while.
- Debug parsing, requires changes to r2 core, unless coredump uses some cool voodoo which I can copy/learn from.

Whats the stance on backwards compatibility for r2pm things as it doesnt seem to use version numbers. i.e. if the latest build of the plugin only works with the latest build of the core will that matter?

**Comment by alvarofe** [comment]
do we have files to test it? the following days i can review this

**Comment by AlexKornitzer** [comment]
There are two full memory dumps for a super simple hello world program (32bit and 64bit) in the regressions `bins/mdmp` folder. Was going to use a mem dump of calc.exe for regressions but it is ~60MB.

Any improvements and thoughts would be greatly welcomed.

**Comment by XVilka** [comment]
I have some MDMP files to test, just privately. Ping me in email.

**Comment by AlexKornitzer** [comment]
Quick little update: modules will now be split into subsections so that we can easily see their contents.

```
iS
[snip]
idx=53 vaddr=0x00400000 paddr=0x00092c8f sz=151552 vsz=151552 perm=m---- name=C:_Users_Developer_Desktop_hello64.exe
idx=54 vaddr=0x00401000 paddr=0x00093c8f sz=7680 vsz=8192 perm=m-r-x name=.text
idx=55 vaddr=0x00403000 paddr=0x00095c8f sz=512 vsz=4096 perm=m-rw- name=.data
idx=56 vaddr=0x00404000 paddr=0x00096c8f sz=2048 vsz=4096 perm=m-r-- name=.rdata
idx=57 vaddr=0x00405000 paddr=0x00097c8f sz=1024 vsz=4096 perm=m-r-- name=.pdata
idx=58 vaddr=0x00406000 paddr=0x00098c8f sz=1024 vsz=4096 perm=m-r-- name=.xdata
idx=59 vaddr=0x00407000 paddr=0x00099c8f sz=0 vsz=4096 perm=m-rw- name=.bss
idx=60 vaddr=0x00408000 paddr=0x0009ac8f sz=2560 vsz=4096 perm=m-rw- name=.idata
idx=61 vaddr=0x00409000 paddr=0x0009bc8f sz=512 vsz=4096 perm=m-rw- name=.CRT
idx=62 vaddr=0x0040a000 paddr=0x0009cc8f sz=512 vsz=4096 perm=m-rw- name=.tls
idx=63 vaddr=0x0040b000 paddr=0x0009dc8f sz=1536 vsz=4096 perm=m-rw- name=.rsrc
idx=64 vaddr=0x0040c000 paddr=0x0009ec8f sz=1536 vsz=4096 perm=m-r-- name=_4
idx=65 vaddr=0x0040d000 paddr=0x0009fc8f sz=57344 vsz=57344 perm=m-r-- name=_19
idx=66 vaddr=0x0041b000 paddr=0x000adc8f sz=7168 vsz=8192 perm=m-r-- name=_31
idx=67 vaddr=0x0041d000 paddr=0x000afc8f sz=6656 vsz=8192 perm=m-r-- name=_45
idx=68 vaddr=0x0041f000 paddr=0x000b1c8f sz=3072 vsz=4096 perm=m-r-- name=_57
idx=69 vaddr=0x00420000 paddr=0x000b2c8f sz=1024 vsz=4096 perm=m-r-- name=_70
idx=70 vaddr=0x00421000 paddr=0x000b3c8f sz=11776 vsz=12288 perm=m-r-- name=_81
idx=71 vaddr=0x00424000 paddr=0x000b6c8f sz=1536 vsz=4096 perm=m-r-- name=_92
idx=72 vaddr=0x77720000 paddr=0x001e2c8f sz=1744896 vsz=1744896 perm=m---- name=C:_Windows_System32_ntdll.dll
idx=73 vaddr=0x77721000 paddr=0x001e3c8f sz=1029632 vsz=1032192 perm=m-r-x name=.text_1
idx=74 vaddr=0x7781d000 paddr=0x002dfc8f sz=512 vsz=4096 perm=m-r-x name=RT
idx=75 vaddr=0x7781e000 paddr=0x002e0c8f sz=190976 vsz=192512 perm=m-r-- name=.rdata_1
idx=76 vaddr=0x7784d000 paddr=0x0030fc8f sz=47104 vsz=57344 perm=m-rw- name=.data_1
idx=77 vaddr=0x7785b000 paddr=0x0031dc8f sz=75776 vsz=77824 perm=m-r-- name=.pdata_1
idx=78 vaddr=0x7786e000 paddr=0x00330c8f sz=369152 vsz=372736 perm=m-r-- name=.rsrc_1
idx=79 vaddr=0x778c9000 paddr=0x0038bc8f sz=1536 vsz=4096 perm=m-r-- name=.reloc
[snip]
```
It looks like I wont be able to remove the memory sections from the Sections view until we modify certain functionality within the RBinPlugin. I am inclined to put forward a case that all vaddr to paddr mapping should be resolved using the RBinMem struct rather than the RBinSection struct. Also there seem to be some relics in the code as `ut64 (*get_vaddr)(RBinFile *arch, ut64 baddr, ut64 paddr, ut64 vaddr)` is never called? Also is `struct r_bin_dbginfo_t *dbginfo` in a functional state or a WIP?

**Pull request merged**